### ISSUE TITLE
feature request: allow counting of empty matches in find/replace form

### ISSUE THREAD

**Issue by molsonkiko** [notepad-plus-plus/notepad-plus-plus#13608]
### Description of the Issue

Empty matches are not counted in the find/replace form. This can be a problem if, for example, you want to [count empty lines in a document](https://community.notepad-plus-plus.org/topic/23951/ignoring-empty-lines-counting/19).

However, I recognize that a user might not want empty matches to be counted, because an empty match is frequently a sign of a poorly written regular expression.

My proposed solution, which I will implement in a PR, is 2-part:
1. Change the counting logic so that empty matches are not excluded.
2.  Change the result displayed by the `Count` button:
  - *If no matches were empty*, it shows something like `Count: 10 matches in entire file` (the old behavior)
  - Otherwise, show something like `Count: 10 matches in entire file (including 3 empty matches)`

### Steps to Reproduce the Issue

1. Create a file with some empty lines:
```
a

b
``` 
3. Use the find/replace form to count occurrences of the regex `^$` (empty line) 
4. The count form will display `0 matches in entire file` even though there is an empty line.

### Expected Behavior

Count empty matches
### Actual Behavior

As described above
### Debug Information

Notepad++ v8.5.1   (64-bit)
Build time : Mar 21 2023 - 19:16:10
Path : C:\Program Files\Notepad++\notepad++.exe
Command Line : 
Admin mode : OFF
Local Conf mode : OFF
Cloud Config : OFF
OS Name : Windows 10 Home (64-bit)
OS Version : 22H2
OS Build : 19045.2846
Current ANSI codepage : 1252
Plugins : 
    ColumnsPlusPlus (0.2.0.6)
    ComparePlus (1.1)
    CSVLint (0.4.6.2)
    CsvQuery (0.1)
    Demo (0.7)
    DSpellCheck (1.5)
    EnhanceAnyLexer (1.1.3)
    HexEditor (0.9.12)
    HugeFiles (0.4.1)
    JsonTools (4.14.0.6)
    mimeTools (2.9)
    NavigateTo (2.4.1.1)
    NppConverter (4.5)
    NppExport (0.4)
    NppLspClient (0.0.13)
    PythonScript (3.0.16)
    QuickText (0.2.5.1)
    RegexTrainer (1)
    XMLTools (3.1.1.13)

**Comment by alankilborn** [comment]
FWIW, a _Find All in Current Document_ search will provide the number of the hits that the _Count_ search does not:

![image](https://user-images.githubusercontent.com/77065706/235906345-945373ae-b463-4a8e-853e-9b4867be4ba8.png)

---------------------------

Interestingly, this type of search doesn't yield a double-click jump to the match point and put up the little calltip box like one would expect, possibly because there is no highlighted match text in the _Search results_ output; here's what I tried:

![image](https://user-images.githubusercontent.com/77065706/235907262-4a0e16e3-e118-4259-a6e7-49752db11d69.png)

But this is fodder for a different issue...

**Comment by alankilborn** [comment]
>Does that mean changing the info-text in the Search Results window as well?

I think that could be avoided. :-)

>Could you please list the exact STR?

Hmm, can't right now as it isn't reproducing.  I have a thought that I initially tried it in a `new 1` file, and I have seen cases where the old problem of double-clicking a line in _Search results_ won't jump to the line in the file,  so maybe that was my case here?

**Comment by molsonkiko** [comment]
> Does that mean changing the info-text in the _Search Results_ window as well?

No, I wasn't planning to do that. The user can see the matches in the _Search Results_ window anyway, so my feature idea isn't needed there.

<details><summary>issue that I can probably figure out on my own</summary>
I've been working on the PR, and while I've gotten the functionality to work pretty well, there's a perplexing inconsistency that I can't understand: empty matches at the start of the file are almost always ignored when there's only one empty match, but if there are multiple empty matches, the empty match at the start of file is always matched. There are no problems that I can see with matching in a selection.

__Example of bad behavior with one empty match at start of file__
![image](https://user-images.githubusercontent.com/46202915/236301320-5d6c26c4-148e-42b5-9fc2-04e1aadad2c0.png)

__Example of appropriate behavior with multiple empty matches, including one at start of file__
![image](https://user-images.githubusercontent.com/46202915/236301126-4fc71612-705d-4741-91df-5494a62307bc.png)
</details>

**Comment by Marmotian** [comment]
In my experience, RegEx searching has some interesting quirks. E.g. searching for "Beginning of line", ^, sometimes includes the last character of the previous line in the match.

In the empty line case, ^$ gets no hits, but ^$. with ". matches newline" ticked gets almost the correct number (it doesn't count the last line of the file if it is blank)

Technically, ^$ shouldn't work for a Windows file because there is always a <CR> before the newline so maybe there's something weird going on with regards to what constitutes a "end of line" that depends of file type?

**Comment by molsonkiko** [comment]
> Technically, ^$ shouldn't work for a Windows file because there is always a {CR} before the newline so maybe there's something weird going on with regards to what constitutes an "end of line" that depends on file type?

That's false. In Windows, the *character sequence* `CR LF` *is* the newline character. Both characters are part of the newline, and in fact Boost regex (the regex engine used by Notepad++) has the option to treat that sequence as a single character.
